File: Graphs/BFS/C++/BFS.cpp
/*
 *	adikul30
 *	Aditya Kulkarni <aditya.kulkarni15@siesgst.ac.in>
 */
#include <bits/stdc++.h>

using namespace std;

#define REP(i, a, b) for(int i = a; i < b; ++i)
#define pb push_back
#define MAX 100

typedef vector<int> vi;

bool visited[MAX];
vi adj[MAX];
int distance[MAX];
queue<int> q;

void bfs(int x){
	visited[x] = true;
	cout << x << endl;
	q.push(x);
	while(!q.empty()){
		int s = q.front();
		q.pop();
		for (auto u : adj[s]){
			if (visited[u]) continue;
			visited[u] = true;
			cout << u << endl;
			q.push(u);
		}
	}
}

int main()
{
	ios_base::sync_with_stdio(0);
	int n = 5;
	REP(i,1,n+1)visited[i] = false;
	adj[1].pb(2);
	adj[1].pb(3);
	adj[1].pb(5);
	adj[2].pb(3);
	adj[3].pb(4);
	
	bfs(1);
	return 0;
}

File: Graphs/BFS/Python/bfs.py
import queue
visited = []
MAX = 20
adj = {1: set([2,3,5]), 2: set([3]), 3: set([8,4]), 4:6}
q = queue.Queue(MAX)

def bfs(x):
	visited.append(x)
	print(x)
	q.put(x)
	while q.empty() == False:
		s = q.get()
		children = adj.get(s)
		if children != None:
			if isinstance(children,int):	# Case : values are not a set
				if children in visited:
					continue
				visited.append(children)
				print(children)
				q.put(children)
			else :							# Case : values are a set
				for i in children:
					if i in visited:
						continue
					visited.append(i)
					print(i)
					q.put(i)


def test():
	global visited
	print("bfs : ")
	visited = []
	bfs(1)

test()